File: accsaber-api/src/main/java/de/ixsen/accsaber/api/controllers/MapLeaderboardsController.java
package de.ixsen.accsaber.api.controllers;

import de.ixsen.accsaber.api.dtos.MapLeaderboardPlayerDto;
import de.ixsen.accsaber.api.mapping.MappingComponent;
import de.ixsen.accsaber.business.ScoreService;
import de.ixsen.accsaber.database.views.AccSaberScore;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("map-leaderboards")
public class MapLeaderboardsController {

    private final MappingComponent mappingComponent;
    private final ScoreService scoreService;

    public MapLeaderboardsController(ScoreService scoreService, MappingComponent mappingComponent) {
        this.scoreService = scoreService;
        this.mappingComponent = mappingComponent;
    }

    @GetMapping("/{mapHash}/{characteristic}/{difficulty}")
    public ResponseEntity<List<MapLeaderboardPlayerDto>> getMapLeaderboards(@PathVariable String mapHash, @PathVariable String characteristic, @PathVariable String difficulty, @RequestParam int page, @RequestParam int pageSize) {
        List<AccSaberScore> scores = this.scoreService.getScoresForMapHash(mapHash, characteristic, difficulty, page, pageSize);
        List<MapLeaderboardPlayerDto> mapLeaderboardPlayerDtos = this.mappingComponent.getMapLeaderboardPlayerMapper().rankedScoresToMapLeaderboardDtos(scores);

        return ResponseEntity.ok(mapLeaderboardPlayerDtos);
    }

    @GetMapping("/{leaderboardId}")
    public ResponseEntity<List<MapLeaderboardPlayerDto>> getMapLeaderboardsByHash(@PathVariable Long leaderboardId) {
        List<AccSaberScore> scores = this.scoreService.getScoresForLeaderboardId(leaderboardId);
        List<MapLeaderboardPlayerDto> mapLeaderboardPlayerDtos = this.mappingComponent.getMapLeaderboardPlayerMapper().rankedScoresToMapLeaderboardDtos(scores);

        return ResponseEntity.ok(mapLeaderboardPlayerDtos);
    }
}

File: accsaber-api/src/main/kotlin/de/ixsen/accsaber/api/controllers/MapLeaderboardsController.kt
package de.ixsen.accsaber.api.controllers

import de.ixsen.accsaber.api.dtos.MapLeaderboardPlayerDto
import de.ixsen.accsaber.api.mapping.MapLeaderboardPlayerMapper
import de.ixsen.accsaber.business.ScoreService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("map-leaderboards")
class MapLeaderboardsController(private val scoreService: ScoreService, private val mapLeaderboardPlayerMapper: MapLeaderboardPlayerMapper) {
    @GetMapping("/{mapHash}/{characteristic}/{difficulty}")
    fun getMapLeaderboards(
        @PathVariable mapHash: String,
        @PathVariable characteristic: String,
        @PathVariable difficulty: String,
        @RequestParam page: Int,
        @RequestParam pageSize: Int
    ): ResponseEntity<List<MapLeaderboardPlayerDto>> {
        val scores = scoreService.getScoresForMapHash(mapHash, characteristic, difficulty, page, pageSize)
        val mapLeaderboardPlayerDtos = this.mapLeaderboardPlayerMapper.rankedScoresToMapLeaderboardDtos(scores)
        return ResponseEntity.ok(mapLeaderboardPlayerDtos)
    }

    @GetMapping("/{mapHash}/{characteristic}/{difficulty}/around/{playerId}")
    fun getMapLeaderboardsAroundPlayer(
        @PathVariable mapHash: String,
        @PathVariable characteristic: String,
        @PathVariable difficulty: String,
        @PathVariable playerId: String
    ) : ResponseEntity<List<MapLeaderboardPlayerDto>> {
        val scores = scoreService.getScoresForMapHashAroundPlayer(mapHash, characteristic, difficulty, playerId)
        val mapLeaderboardPlayerDtos = this.mapLeaderboardPlayerMapper.rankedScoresToMapLeaderboardDtos(scores)
        return ResponseEntity.ok(mapLeaderboardPlayerDtos)
    }

    @GetMapping("/{leaderboardId}")
    fun getMapLeaderboardsByHash(@PathVariable leaderboardId: Long): ResponseEntity<List<MapLeaderboardPlayerDto>> {
        val scores = scoreService.getScoresForLeaderboardId(leaderboardId)
        val mapLeaderboardPlayerDtos = this.mapLeaderboardPlayerMapper.rankedScoresToMapLeaderboardDtos(scores)
        return ResponseEntity.ok(mapLeaderboardPlayerDtos)
    }
}